content/library/advanced-features/custom-classes.md

```diff
@@ -0,0 +1,150 @@
+
+1. Defining a custom `@dataclass` to store data within your app
+2. Defining an `Enum` class to represent a fixed set of options.
+3. Defining custom interfaces to external services or databases not covered by Streamlit's `experimental_connection` [API](/library/api-reference/connections/st.experimental_connection).
+
+Because streamlit re-runs your main script from top to bottom after every user interaction, care must be taken when defining custom classes to ensure that a re-definition of the class type during these re-runs will not cause unwanted effects. By following the guidelines on this page, you will be less likely to encounter common pitfalls.
+
+<Collapse title="Table of contents" expanded={true}>
+
+1. [General guildelines](#general-guidelines)
+2. [Enums](#enums)
+
+</Collapse>
+
+## General Guidelines
+
+**For all classes:**
+
+- If possible, move class definitions into their own module file.
+
+  For example:
+
+  ```python
+  # In Main.py ---------------------------------------
+  import streamlit as st
+
+  # MyClass gets re-defined every time Main.py re-runs
+  class MyClass:
+    def __init__(self, var1, var2):
+        ...
+
+  st.write(MyClass("foo", "bar"))
+  ```
+
+  becomes:
+
+  ```python
+  # In my_class.py ---------------------------------------
+  class MyClass:
+    def __init__(self, var1, var2):
+        ...
+
+  # In Main.py -------------------------------------------
+  import streamlit as st
+  from my_class import MyClass
+
+  # MyClass doesn't change identity when Main.py re-runs
+
+  st.write(MyClass("foo", "bar"))
+  ```
+
+  Streamlit only reloads code in imported modules when it detects the code has changed. Thus, each time your page script re-runs it is using the exact same `MyClass` as on the previous execution.
+
+**For classes that store data (like [dataclasses](https://docs.python.org/3/library/dataclasses.html)):**
+
+- Consider defining a custom `__eq__` method based on the data _in_ the class.
+
+  ```python
+  import streamlit as st
+  from dataclasses import dataclass
+
+  @dataclass
+  class MyDataclass:
+    var1: int
+    var2: float
+
+    def __eq__(self, other):
+      # Two instances of MyDataclass are equal if both of their
+      # fields match.
+      return (self.var1, self.var2) == (other.var1, other.var2)
+
+  st.write(MyDataclass(1, 5.5) == MyDataclass(1, 5.5))
+  ```
+
+  The default python `__eq__` implementations for both regular
+  classes as well as `@dataclasses` depend on the in-memory ID of the class or class instance. This can cause comparison problems where two
+  dataclass instances that you think should be equal are not evaluated
+  as being equal.
+
+- Alternatively, if you are storing data in `st.session_state`,
+  consider defining serialization and deserialization methods like `to_str` and `from_str` for your class, and use these to store class instance data in `st.session_state` rather than storing the class instance itself.
```

Comment:
Can you explain why this is necessary or helpful? It might be, but I'm not sure when, as session_state works perfectly well to store whole dataclass objects. If you do need to serialize, then I would bias towards using the `asdict` or `astuple` methods from the dataclass module.

Reply:
Yeah, I am working to see if I can explain the problem a little better. Basically if you do something like 
```
@dataclass
class MyDataclass:
   value:  str

if "foo" in st.session_state and st.session_state["foo"] == MyDataclass("bar"):
   st.success("you win")
if st.button("click me"):
   st.session_state["foo"] = MyDataclass("bar")
```
you'll never see that `st.success` because session_state has an old instance of MyDataclass in it. The suggestion to add a serialization method is only if you don't want to define a custom `__eq__` or put your class in a separate module file, hence why it's marked as "Alternatively";

As a note: even putting the class in a separate module file is not a fool-proof way out of this problem, it only reduces the frequency with which it can occur to "times when streamlit reloads modules, e.g. when someone edits a bit of page code". So the problem can still pop up while you're developing pages if the only mitigation you took is moving the class into a separate file. @sfc-gh-dmatthews do you think it makes sense to mention that in here or is that going too deep?

Reply:
I definitely think it's worth mentioning as "during development" is exactly when people are going to be encountering this and getting confused.